controller/system_restore_controller.go

```diff
@@ -0,0 +1,397 @@
+	log := c.getLoggerForSystemRestore(systemRestore.Name)
+
+	systemBackup, err := c.ds.GetSystemBackupRO(systemRestore.Spec.SystemBackup)
+	if err != nil {
+		return err
+	}
+
+	cfg, err := backupTargetClient.SystemBackupConfigGet(systemBackup.Name, systemBackup.Status.Version)
+	if err != nil {
+		return err
+	}
+
+	serviceAccountName, err := c.getLonghornServiceAccountName()
+	if err != nil {
+		return err
+	}
+
+	job, err := c.ds.CreateJob(c.newSystemRestoreJob(systemRestore.Name, c.namespace, cfg.ManagerImage, serviceAccountName))
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Created system restore job %v", job.Name)
+
+	return nil
+}
+
+func (c *SystemRestoreController) newSystemRestoreJob(systemRestoreName, namespace, managerImage, serviceAccount string) *batchv1.Job {
+	backoffLimit := int32(RestoreJobBackoffLimit)
+	privileged := true
+
+	cmd := []string{
+		"longhorn-manager", "system-rollout", systemRestoreName,
+	}
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      systemRestoreName,
+			Namespace: namespace,
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit: &backoffLimit,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: systemRestoreName,
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: serviceAccount,
+					Containers: []corev1.Container{
+						{
+							Name:    systemRestoreName,
+							Image:   managerImage,
+							Command: cmd,
+							Env: []corev1.EnvVar{
+								{
+									Name: "POD_NAMESPACE",
+									ValueFrom: &corev1.EnvVarSource{
+										FieldRef: &corev1.ObjectFieldSelector{
+											FieldPath: "metadata.namespace",
+										},
+									},
+								},
+								{
+									Name:  "NODE_NAME",
+									Value: c.controllerID,
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "engine",
+									MountPath: types.EngineBinaryDirectoryOnHost,
+								},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Privileged: &privileged,
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: "engine",
+							VolumeSource: corev1.VolumeSource{
+								HostPath: &corev1.HostPathVolumeSource{
+									Path: types.EngineBinaryDirectoryOnHost,
+								},
+							},
+						},
+					},
+					RestartPolicy: corev1.RestartPolicyOnFailure,
+					NodeSelector: map[string]string{
+						corev1.LabelHostname: c.controllerID,
+					},
+				},
+			},
+		},
+	}
+
+	return job
+}
+
+func (c *SystemRestoreController) getLonghornServiceAccountName() (string, error) {
+	managerPods, err := c.ds.ListManagerPods()
+	if err != nil {
+		return "", err
+	}
+
+	for _, pod := range managerPods {
+		return pod.Spec.ServiceAccountName, nil
+	}
+	return "", errors.Errorf("failed to find service account from manager pods")
+}
```

Comment:
Move to `VolumeManager`?

Reply:
For any reason..?

Reply:
It seems IDE messed up the comment. Let's ignore this.